### PR TITLE
Deploy Attacher as container in  provisioner statefulset

### PR DESCRIPTION
**Pull request by fs-system**
Deploy attacher as  a container in provisioner statefulset instead  of deploying it as a separate statefulset

This PR also adds functionality to create CSIDriver object which will  help people to list the CSI drivers deployed in theier environment

### CODE REVIEW

File: templates/csi-driver-object.yaml.j2 [hunk 1]
```diff
@@ -0,0 +1,7 @@
+apiVersion: storage.k8s.io/v1beta1
+kind: CSIDriver
+metadata:
+  name: kadalu.gluster
```
Comment (amarts) [inline templates/csi-driver-object.yaml.j2]: if we merge #58 this needs to change too.

**Review by amarts — COMMENTED** [review]
lgtm

File: tests/minikube.sh [hunk 2]
```diff
@@ -105,20 +106,20 @@ kadalu_operator)
     # give it some time
     cnt=0
     while true; do
-        cnt=$((cnt+1))
-        sleep 1;
-        ret=`kubectl get pods -nkadalu -o name | wc -l`
+        cnt=$((cnt + 1))
+        sleep 1
+        ret=$(kubectl get pods -nkadalu -o name | wc -l)
         if [[ $ret -ge 5 ]]; then
```
Comment (amarts) [inline tests/minikube.sh]: this needs to change as a test for 4 pods, not 5 like now.

File: operator/main.py [hunk 3]
```diff
@@ -293,6 +293,16 @@ def deploy_csi_pods(core_v1_client):
             return
 
     # Deploy CSI Pods
+    api_instance=client.VersionApi().get_code()
+    if api_instance.major>"1" or api_instance.major=="1" and api_instance.minor>="14":
+        filename = os.path.join(MANIFESTS_DIR, "csi-driver-object.yaml")
+        template(filename, namespace=NAMESPACE, kadalu_version=VERSION)
+        execute(KUBECTL_CMD, "create", "-f", filename)
+    else:
```
Comment (fs-system) [inline operator/main.py:301]: btw why kadalu require minimum  1.14  as stated here? https://github.com/kadalu/kadalu#pre-requisites

@amarts

Comment (amarts) [reply operator/main.py:301]: that requirement came from what we tested.... We haven't tested for version below that.

Comment (fs-system) [reply operator/main.py:301]: if the minimum version is 1.14 i can remove if else check

**Comment by aravindavk** [comment]
Looks good. Thanks @fs-system 

Once CI issues resolved and comments are addressed then we can merge this one.

**Comment by fs-system** [comment]
tests looks flaky https://travis-ci.org/kadalu/kadalu/jobs/612282963#L1583

pods are coming up locally

```
[$]kubectl logs po/operator-8576ccccf9-5mcvn  -nkadalu
[2019-11-15 09:17:24,961] DEBUG [rest - 219:request] - response body: {"kind":"ConfigMapList","apiVersion":"v1","metadata":{"selfLink":"/api/v1/namespaces/kadalu/configmaps","resourceVersion":"1039"},"items":[]}
[2019-11-15 09:17:27,591] INFO [main - 331:deploy_config_map] - Deployed ConfigMap	 manifest=/kadalu/templates/configmap.yaml
[2019-11-15 09:17:27,598] DEBUG [rest - 219:request] - response body: {"kind":"PodList","apiVersion":"v1","metadata":{"selfLink":"/api/v1/namespaces/kadalu/pods","resourceVersion":"1045"},"items":[{"metadata":{"name":"operator-8576ccccf9-5mcvn","generateName":"operator-8576ccccf9-","namespace":"kadalu","selfLink":"/api/v1/namespaces/kadalu/pods/operator-8576ccccf9-5mcvn","uid":"eec6f547-11a9-403a-a0fd-f56eaf83e36d","resourceVersion":"1040","creationTimestamp":"2019-11-15T09:16:34Z","labels":{"name":"kadalu","pod-template-hash":"8576ccccf9"},"ownerReferences":[{"apiVersion":"apps/v1","kind":"ReplicaSet","name":"operator-8576ccccf9","uid":"8f031fbb-5d8f-4647-adf5-62e8f5019d1f","controller":true,"blockOwnerDeletion":true}]},"spec":{"volumes":[{"name":"kadalu-operator-sa-token-sv99x","secret":{"secretName":"kadalu-operator-sa-token-sv99x","defaultMode":420}}],"containers":[{"name":"kadalu-operator","image":"madhupr001/test","env":[{"name":"WATCH_NAMESPACE","valueFrom":{"fieldRef":{"apiVersion":"v1","fieldPath":"metadata.namespace"}}},{"name":"POD_NAME","valueFrom":{"fieldRef":{"apiVersion":"v1","fieldPath":"metadata.name"}}},{"name":"OPERATOR_NAME","value":"kadalu-operator"},{"name":"DOCKER_USER","value":"kadalu"},{"name":"KADALU_VERSION","value":"latest"}],"resources":{},"volumeMounts":[{"name":"kadalu-operator-sa-token-sv99x","readOnly":true,"mountPath":"/var/run/secrets/kubernetes.io/serviceaccount"}],"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"Always","securityContext":{"capabilities":{},"privileged":true}}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","serviceAccountName":"kadalu-operator-sa","serviceAccount":"kadalu-operator-sa","nodeName":"worker0","securityContext":{},"schedulerName":"default-scheduler","tolerations":[{"key":"node.kubernetes.io/not-ready","operator":"Exists","effect":"NoExecute","tolerationSeconds":300},{"key":"node.kubernetes.io/unreachable","operator":"Exists","effect":"NoExecute","tolerationSeconds":300}],"priority":0,"enableServiceLinks":true},"status":{"phase":"Running","conditions":[{"type":"Initialized","status":"True","lastProbeTime":null,"lastTransitionTime":"2019-11-15T09:16:34Z"},{"type":"Ready","status":"True","lastProbeTime":null,"lastTransitionTime":"2019-11-15T09:17:25Z"},{"type":"ContainersReady","status":"True","lastProbeTime":null,"lastTransitionTime":"2019-11-15T09:17:25Z"},{"type":"PodScheduled","status":"True","lastProbeTime":null,"lastTransitionTime":"2019-11-15T09:16:34Z"}],"hostIP":"192.168.121.110","podIP":"10.44.0.1","startTime":"2019-11-15T09:16:34Z","containerStatuses":[{"name":"kadalu-operator","state":{"running":{"startedAt":"2019-11-15T09:17:24Z"}},"lastState":{},"ready":true,"restartCount":0,"image":"docker.io/madhupr001/test:latest","imageID":"docker-pullable://docker.io/fs-system/test@sha256:d7c341cc963e7b5980e6bf25453d0a95ab62e7689240a4e3813a64e978af4572","containerID":"docker://3efa77904ae1385c7ae82398e19efc3e997c821c1a33280523fb5bcf4e725678"}],"qosClass":"BestEffort"}}]}
[2019-11-15 09:17:27,626] DEBUG [rest - 219:request] - response body: {
  "major": "1",
  "minor": "15",
  "gitVersion": "v1.15.0",
  "gitCommit": "e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529",
  "gitTreeState": "clean",
  "buildDate": "2019-06-19T16:32:14Z",
  "goVersion": "go1.12.5",
  "compiler": "gc",
  "platform": "linux/amd64"
}
[2019-11-15 09:17:28,590] INFO [main - 311:deploy_csi_pods] - Deployed CSI Pods	 manifest=/kadalu/templates/csi.yaml
[2019-11-15 09:17:28,624] DEBUG [rest - 219:request] - response body: {"kind":"StorageClassList","apiVersion":"storage.k8s.io/v1","metadata":{"selfLink":"/apis/storage.k8s.io/v1/storageclasses","resourceVersion":"1060"},"items":[]}
[2019-11-15 09:17:28,976] INFO [main - 347:deploy_storage_class] - Deployed StorageClass	 manifest=/kadalu/templates/storageclass.yaml
[ $]kubectl get po  -nkadalu
NAME                        READY   STATUS              RESTARTS   AGE
csi-nodeplugin-bcmsl        0/2     ContainerCreating   0          44s
csi-provisioner-0           0/3     ContainerCreating   0          44s
operator-8576ccccf9-5mcvn   1/1     Running             0          98s
[ $]kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.2", GitCommit:"66049e3b21efe110454d67df4fa62b08ea79a19b", GitTreeState:"clean", BuildDate:"2019-05-16T16:23:09Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-19T16:32:14Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
[$]kubectl get po  -nkadalu
NAME                        READY   STATUS    RESTARTS   AGE
csi-nodeplugin-bcmsl        2/2     Running   0          70s
csi-provisioner-0           3/3     Running   0          70s
operator-8576ccccf9-5mcvn   1/1     Running   0          2m4s
```

**Comment by fs-system** [comment]
@aravindavk @amarts can you please retrigger the CI?

**Comment by amarts** [comment]
@fs-system Please note that with the latest changes, the tests on 1.13.7 passed all fine, but it failed on 1.15.0 --- https://travis-ci.org/kadalu/kadalu/jobs/612283621?utm_medium=notification&utm_source=github_status

Can you check? also now that we have 1.16.0 available, it is fine to test with one lowest (say 1.13.7), and one latest, so you can propose 1.16.0 there too.

File: operator/main.py (hunk 3 above)
Comment (amarts) [reply operator/main.py:301]: @aravindavk you have any idea about why it is done this way? different between object / crd yml?

File: manifests/kadalu-operator.yaml [hunk 4]
```diff
@@ -149,7 +141,10 @@ rules:
     verbs: ["create", "list", "watch", "delete"]
   - apiGroups: ["csi.storage.k8s.io"]
     resources: ["csidrivers"]
-    verbs: ["create", "delete"]
+    verbs: ["get", "create", "delete"]
+  - apiGroups: ["storage.k8s.io"]
```
Comment (aravindavk) [inline manifests/kadalu-operator.yaml:147]: Move these changes to `operator.yaml.j2` and use `make gen-manifest` to generate this file

File: operator/main.py (hunk 3 above)
Comment (fs-system) [reply operator/main.py:301]: @amarts more info here https://kubernetes-csi.github.io/docs/csi-driver-object.html

**Comment by amarts** [comment]
Planning to separate changes to .travis.yml to add 1.16.0 k8s as a separate patch. Let me see what we get there.

**Comment by amarts** [comment]
Closed with #92

**Pull request closed without merging**